core/src/main/scala/kafka/admin/TopicCommand.scala

```diff
@@ -634,7 +639,8 @@ object TopicCommand extends Logging {
       if (!has(listOpt) && !has(describeOpt))
         CommandLineUtils.checkRequiredArgs(parser, options, topicOpt)
       if (has(createOpt) && !has(replicaAssignmentOpt))
-        CommandLineUtils.checkRequiredArgs(parser, options, partitionsOpt, replicationFactorOpt)
+        if (has(partitionsOpt) || has(replicationFactorOpt))
+          CommandLineUtils.checkRequiredArgs(parser, options, partitionsOpt, replicationFactorOpt)
```

Comment:
Not sure I get the motivation for this. Seems useful to be able to specify `--partitions` only and rely on the default replication factor. I guess this goes back to the fact that we don't have a convenient constructor for `NewTopic` when only one of the options is provided? I wonder if it makes sense to add a constructor like the following:
```java
public NewTopic(String name, Optional<Integer> numPartitions, Optional<Short> replicationFactor)
```

Reply:
You'll see a message in the KIP discussion suggesting exactly this. Can you guess who sent it?

Reply:
Hehe you this discussion was discussed heavily in the KIP. The idea is to introduce a follow-up KIP containing a `NewTopic#Builder` that can have whatever amazing combo of parameters that you want.

Reply:
But if we're moving to a builder (which makes sense), then we'd just want a single usable constructor rather than a bunch of different options... I'd probably drop the `NewTopic(String)` constructor in favor of the one above. Otherwise users have to use the other constructor and the -1 sentinel to allow for optional defaults. Putting that aside, from the perspective of the command though, is there any good reason to constrain the options?

Reply:
I think there's no harm in the builder using the constructor with sentinel values internally (the user wouldn't see this). That way the user has three options: 
- name only constructor (e.g. `new NewTopic(name)`)
- builder for either partitions or replicas (e.g. `new Builder(name).partitions(1).build()`)
- name, partitions and replicas constructor (e.g. `new NewTopic(name, 1, (short) 1)`) - honestly I'd deprecate this one in favor of the builder, but that's just personal preference.

Since I feel like it would be more common to have either both or neither, the `String, Optional, Optional` constructor would be adding bloat to optimize for the less common cases (and I don't think we should have both `String` and `String, Optional, Optional`) such as `new NewTopic(name, Optional.empty(), Optional.empty())`.

That being said, I don't feel so strongly about this - the only reason I didn't include this in the PR is that it seemed to be contentious in the KIP discussion. Also with regards to the cli - I'm happy to allow users to specify one or the other, I just thought it's an inconsistent user experience to have it be possible to do something from the CLI and not from the AdminClient (though, again, this is just an opinion and I'm happy to allow this).

Let me know what you think :)

Reply:
It depends on how you look at it. Every topic CLI will use the constructor with the optionals, if available. Programmatic topic creation would probably just use the String one.

Reply:
Discussed this offline. The argument I was making is that we don't have the builder now and we may never have it if no one follows up with a KIP. We should make sure that this KIP leaves us in a good state without relying on future work. I would prefer either adding the builder into this KIP or giving users a type-safe way to specify the options they want. In any case, we definitely don't want to be in a state where we have arbitrary restrictions because we couldn't decide what color to paint the shed.

Reply:
Consider me convinced - I'll add the Optional constructor in this PR.